### PR TITLE
Always render all sides in the overlay

### PR DESCRIPTION
**Pull request by Pokechu22**
This change makes it so that the overlay always renders on all sides, regardless of what the corresponding block returns for `shouldSideBeRendered`.

## Why?

Here's some situations where the current behavior can cause problems (both have affected me):

### Scenario 1

A player attempts to build a water elevator from a schematic.  They build the outside, and then start filling it with water.  However, they accidently miss one source block.  While the schematic looks fine (none of the water blocks are highlighted orange to indicate the wrong data value and thus a lack of a source), there still is a slow spot in the elevator due to the missing source.

<table>
<tr><th>Current</th><th>With this fix</th></tr>
<tr><td>Water elevator from inside that looks fine but isn't</td><td>Water elevator from inside with the problem clearly highlighted</td></tr>
<tr><th colspan="2">Explanation</th></tr>
<tr><td colspan="2">Water elevator from outside - problem visible in both cases, but you can't always get outside</td></tr>
</table>

This happens because liquids don't render sides next to others of the same liquid.

### Scenario 2

A player is building a circuit involving comparators and hoppers.  Unfortunately, they forget a few comparators.  The circuit doesn't work, even though it looks perfect from above, as schematica renders a _fake_ comparator where the real one should be, without highlighting it.  (Note that this is far worse when the comparator is surrounded by blocks on all sides, as you don't even get the vertical sides).

<table>
<tr><th>Current</th><th>With this fix</th></tr>
<tr><td>Circuit that looks perfectly fine except for a few blue streaks</td><td>Same circuit, but it's now clear that there are missing comparators</td></tr>
</table>

Or, they brainfart and place repeaters where some of the comparators should be:

<table>
<tr><th>Current</th><th>With this fix</th></tr>
<tr><td>Derpy circuit, with a few streaks of red on repeaters but mostly looking (incorrectly) fine</td><td>Same circuit now clearly showing that the repeaters shouldn't be</td></tr>
</table>

Both of these happen because repeaters and comparators return false in `shouldSideBeRendered` for vertical sides.

## A warning

I do note that #109 was fixed by _adding_ more side restrictions, so this change may reintroduce that bug.  However, I haven't experienced it, even with large areas of repeaters.

### CODE REVIEW

**Comment by Lunatrius** [comment]
I'll check if I can find the schematic that I used to reproduce the issue. I'll also make sure to check if the vanilla code where this crash happened changed.

**Comment by Pokechu22** [comment]
Thanks.  If you have the original crash report, can you send it to me so that I can also check?

**Comment by Lunatrius** [comment]
Found this old crash report. After digging into the 1.12 codebase I don't see any reason why it would break.

```
---- Minecraft Crash Report ----
// Why is it breaking :(

Time: 12/26/15 9:47 AM
Description: Batching chunks

java.lang.IndexOutOfBoundsException
	at java.nio.Buffer.checkIndex(Buffer.java:546)
	at java.nio.DirectByteBuffer.putFloat(DirectByteBuffer.java:893)
	at net.minecraft.client.renderer.WorldRenderer.func_181662_b(WorldRenderer.java:437)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawQuads(GeometryTessellator.java:141)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawQuads(GeometryTessellator.java:108)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawCuboid(GeometryTessellator.java:90)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawCuboid(GeometryTessellator.java:73)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawCuboid(GeometryTessellator.java:69)
	at com.github.lunatrius.schematica.client.renderer.chunk.overlay.RenderOverlay.func_178581_b(RenderOverlay.java:157)
	at net.minecraft.client.renderer.chunk.ChunkRenderWorker.func_178474_a(SourceFile:78)
	at net.minecraft.client.renderer.chunk.ChunkRenderWorker.run(SourceFile:38)
	at java.lang.Thread.run(Thread.java:745)

A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Stacktrace:
	at java.nio.Buffer.checkIndex(Buffer.java:546)
	at java.nio.DirectByteBuffer.putFloat(DirectByteBuffer.java:893)
	at net.minecraft.client.renderer.WorldRenderer.func_181662_b(WorldRenderer.java:437)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawQuads(GeometryTessellator.java:141)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawQuads(GeometryTessellator.java:108)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawCuboid(GeometryTessellator.java:90)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawCuboid(GeometryTessellator.java:73)
	at com.github.lunatrius.core.client.renderer.GeometryTessellator.drawCuboid(GeometryTessellator.java:69)

-- Affected level --
Details:
Level name: MpServer
All players: 1 total; [EntityPlayerSP['Maverick25'/30205, l='MpServer', x=163.46, y=84.99, z=-185.91]]
Chunk stats: MultiplayerChunkCache: 430, 430
Level seed: 0
Level generator: ID 00 - default, ver 1. Features enabled: false
Level generator options:
Level spawn location: 236.00,64.00,252.00 - World: (236,64,252), Chunk: (at 12,4,12 in 14,15; contains blocks 224,0,240 to 239,255,255), Region: (0,0; contains chunks 0,0 to 31,31, blocks 0,0,0 to 511,255,511)
Level time: 55563 game time, 55563 day time
Level dimension: 0
Level storage version: 0x00000 - Unknown?
Level weather: Rain time: 0 (now: false), thunder time: 0 (now: false)
Level game mode: Game mode: creative (ID 1). Hardcore: false. Cheats: false
Forced entities: 22 total; [EntityPlayerSP['Maverick25'/30205, l='MpServer', x=163.46, y=84.99, z=-185.91], EntityPig['Pig'/36101, l='MpServer', x=195.56, y=63.00, z=-209.56], EntityPig['Pig'/36102, l='MpServer', x=193.78, y=63.00, z=-218.28], EntityPig['Pig'/36103, l='MpServer', x=195.06, y=63.00, z=-203.50], EntityPig['Pig'/36104, l='MpServer', x=205.47, y=63.00, z=-192.88], EntityPig['Pig'/36105, l='MpServer', x=197.06, y=63.00, z=-207.81], EntityPig['Pig'/36106, l='MpServer', x=197.56, y=63.00, z=-173.47], EntityPig['Pig'/36107, l='MpServer', x=196.78, y=63.00, z=-174.69], EntityPig['Pig'/36108, l='MpServer', x=193.16, y=63.00, z=-174.47], EntityPig['Pig'/36109, l='MpServer', x=192.50, y=72.00, z=-171.50], EntityPig['Pig'/36111, l='MpServer', x=196.50, y=70.00, z=-139.50], EntityZombie['Zombie'/36636, l='MpServer', x=141.50, y=14.00, z=-154.50], EntityZombie['Zombie'/36637, l='MpServer', x=143.50, y=14.00, z=-156.50], EntityItem['item.tile.doublePlant.syringa'/36904, l='MpServer', x=149.68, y=64.00, z=-170.61], EntitySkeleton['Skeleton'/36937, l='MpServer', x=143.50, y=22.00, z=-139.50], EntitySkeleton['Skeleton'/36938, l='MpServer', x=142.50, y=22.00, z=-140.50], EntitySkeleton['Skeleton'/36939, l='MpServer', x=142.50, y=22.00, z=-142.50], EntityItem['item.tile.flower1.dandelion'/35818, l='MpServer', x=143.69, y=48.00, z=-191.22], EntityPig['Pig'/35951, l='MpServer', x=188.63, y=63.00, z=-216.84], EntityPig['Pig'/35952, l='MpServer', x=176.09, y=63.00, z=-221.38], EntityPig['Pig'/35953, l='MpServer', x=186.50, y=63.00, z=-216.50], EntityPig['Pig'/35954, l='MpServer', x=191.50, y=63.00, z=-207.50]]
Retry entities: 0 total; []
Server brand: Spigot
Server type: Non-integrated multiplayer server
Stacktrace:
	at net.minecraft.client.multiplayer.WorldClient.func_72914_a(WorldClient.java:347)
	at net.minecraft.client.Minecraft.func_71396_d(Minecraft.java:2536)
	at net.minecraft.client.renderer.chunk.ChunkRenderWorker.run(SourceFile:44)
	at java.lang.Thread.run(Thread.java:745)

-- System Details --
Details:
Minecraft Version: 1.8.8
Operating System: Windows 10 (amd64) version 10.0
Java Version: 1.8.0_25, Oracle Corporation
Java VM Version: Java HotSpot(TM) 64-Bit Server VM (mixed mode), Oracle Corporation
Memory: 309950120 bytes (295 MB) / 653901824 bytes (623 MB) up to 1060372480 bytes (1011 MB)
JVM Flags: 6 total; -XX:HeapDumpPath=MojangTricksIntelDriversForPerformance_javaw.exe_minecraft.exe.heapdump -Xmx1G -XX:+UseConcMarkSweepGC -XX:+CMSIncrementalMode -XX:-UseAdaptiveSizePolicy -Xmn128M
IntCache: cache: 0, tcache: 0, allocated: 13, tallocated: 95
FML: MCP 9.18 Powered by Forge 11.15.0.1647 5 mods loaded, 5 mods active
States: 'U' = Unloaded 'L' = Loaded 'C' = Constructed 'H' = Pre-initialized 'I' = Initialized 'J' = Post-initialized 'A' = Available 'D' = Disabled 'E' = Errored
	UCHIJA mcp{9.18} [Minecraft Coder Pack] (minecraft.jar)
	UCHIJA FML{8.0.99.99} [Forge Mod Loader] (forge-1.8.8-11.15.0.1647-1.8.8.jar)
	UCHIJA Forge{11.15.0.1647} [Minecraft Forge] (forge-1.8.8-11.15.0.1647-1.8.8.jar)
	UCHIJA LunatriusCore{1.1.2.31} [LunatriusCore] (LunatriusCore-1.8.8-1.1.2.31-universal.jar)
	UCHIJA Schematica{1.7.7.141} [Schematica] (Schematica-1.8.8-1.7.7.141-universal.jar)
Loaded coremods (and transformers):
GL info: ~~ERROR~~ RuntimeException: No OpenGL context found in the current thread.
Launched Version: 1.8.8-forge1.8.8-11.15.0.1647-1.8.8
LWJGL: 2.9.4
OpenGL: ~~ERROR~~ RuntimeException: No OpenGL context found in the current thread.
GL Caps: Using GL 1.3 multitexturing.
Using GL 1.3 texture combiners.
Using framebuffer objects because OpenGL 3.0 is supported and separate blending is supported.
Shaders are available because OpenGL 2.1 is supported.
VBOs are available because OpenGL 1.5 is supported.

Using VBOs: Yes
Is Modded: Definitely; Client brand changed to 'fml,forge'
Type: Client (map_client.txt)
Resource Packs:
Current Language: English (US)
Profiler Position: N/A (disabled)
CPU: 6x AMD FX(tm)-6300 Six-Core Processor
```

**Comment by Lunatrius** [comment]
Alternatively I could disable the check for any blocks that aren't full sized solid blocks. I'll play around with this later today to see if I can figure out something.

**Comment by Pokechu22** [comment]
I can confirm that this was a vanilla bug ([MC-81738](https://bugs.mojang.com/browse/MC-81738 "[MC-81738] Crash: IndexOutOfBoundsException on Tesselating block in world") that was fixed in 15w38a (&rarr; 1.9).  While the crash reports for that are 1.8.7, I can confirm that `bfd` is `WorldRenderer` in both 1.8.7 and 1.8.8, and it's `func_181662_b` in both cases (same line numbers).

So I _think_ everything should be fine.  I still don't fully understand this part of the code, though.

**Comment by Lunatrius** [comment]
That method hides all invisible faces. This makes perfect sense for solid blocks, but might require some special tweaking or exclusion from the filtering.

**Comment by Pokechu22** [comment]
Alright, I understand it somewhat, and have found a drawback to this (and another benefit).

This is the problem with this change (which is what you don't like):

<table>
<tr><th colspan="2">From outside</th></tr>
<tr><td colspan="2">From outside: a large stone brick cube</td></tr>
<tr><th>Normal inside</th><th>Inside with this change</th></tr>
<tr><td>Inside, normally</td><td>Inside, sO mUcH bLuE</td></tr>
</table>

That's a lot of uneeded polygons, and isn't that helpful either.

-----

On the other hand, here's another benefit:

<table>
<tr><th>The setup</th></tr>
<tr><td>Outside</td></tr>
<tr><th>Without this change</th></tr>
<tr><td>Without this change: no line visible between two blocks indicating a missing one</td></tr>
<tr><th>With this change</th></tr>
<tr><td>With this change: a faint line is visible</td></tr>

It's more visible full-screen, but there's a small indication of that there are missing blocks behind.  I actually find this fairly useful.  This same indication is sometimes visible in layers mode and near the edges of schematics, but not always (and that makes it not usable).

-----

Perhaps the solution is to completely ignore the block type; instead the sides should be rendered based off of just whether or not there is a missing block in comparison to the schematic.  I'm not entirely sure what that'd look like, but I think it would allow for more accuracy (including the corners) without rendering everything in solid blocks.

**Comment by Lunatrius** [comment]
I'm not entirely sure what the best solution to this is. Removing the check entirely makes a lot of the cases look nice, but it inside of solid objects there's simply too much rendering.

One way to cut down the rendering inside of solid blocks would be to check the surrounding 3x3x3 area for any non solids and skip the overlay if all 9 are solid. It fixes the majority, but it will look weird around the edges as one "layer" of the overlay will always render inside.

**Comment by Lunatrius** [comment]
*Random shower thought ahead.*

Right now the renderer decides which side to render based on the block. I could add neighbor check, but that would unnecessarily slow down everything. Instead I could keep a 3 dimensional array inside the schematic that keeps track of the block "states" (matches world, missing, wrong block, wrong meta etc).

With that information cached I could simply query the information for a block and render the sides only when the adjacent block has a different value. I created a simple 2D example.

image

**Comment by Pokechu22** [comment]
Yea, that's what I was thinking too.  I'm going to try to implement that.

**Comment by Pokechu22** [comment]
OK, I've tried to implement this.  It's slightly different than what you suggested; I instead generate the array in `RenderOverlay.rebuildChunk` (and the array is _only_ for the chunk section being rendered).  This works fairly well, with one problem: the overlay is always rendered at chunk section boundaries (every 16 blocks) even when it doesn't need to be.

With the approach you suggested of having it in the schematic, that section issue could be avoided, but I'm not entirely sure how I'd go about implementing it there (making sure all the data is up-to-date).

On second thought, I could avoid the problem fairly easily - I just need to go out one block further on each edge...

**Comment by Pokechu22** [comment]
Done; works nice now.  (I've also switched to `MutableBlockPos` which should offer a small performance improvement)

File: src/main/java/com/github/lunatrius/schematica/client/renderer/chunk/overlay/RenderOverlay.java [hunk 1]
```diff
@@ -146,34 +165,45 @@ public void rebuildChunk(final float x, final float y, final float z, final Chun
         compiledOverlay.setVisibility(visgraph.computeVisibility());
     }
 
-    private int getSides(final IBlockState blockState, final Block block, final World world, final BlockPos pos, int sides) {
-        if (block.shouldSideBeRendered(blockState, world, pos, EnumFacing.DOWN)) {
+    private int getSides(final BlockType[][][] types, final int x, final int y, final int z) {
+        // The padding cannot be rendered (it lacks neighbors)
+        assert x > 0 && x < 17;
```
Comment (Lunatrius) [inline src/main/java/com/github/lunatrius/schematica/client/renderer/chunk/overlay/RenderOverlay.java]: Are these not compiled into the following code, which is not really what we'd want?
```java
if (!(x > 0 && x < 17))
    throw new AssertionError();
```

File: src/main/java/com/github/lunatrius/schematica/client/renderer/chunk/overlay/RenderOverlay.java [hunk 2]
```diff
@@ -146,34 +165,45 @@ public void rebuildChunk(final float x, final float y, final float z, final Chun
...
 
-        if (block.shouldSideBeRendered(blockState, world, pos, EnumFacing.EAST)) {
+        if (types[x + 1][y][z] != type) {
             sides |= GeometryMasks.Quad.EAST;
         }
 
         return sides;
     }
 
+
```
Comment (Lunatrius) [inline src/main/java/com/github/lunatrius/schematica/client/renderer/chunk/overlay/RenderOverlay.java]: *Noise.*

File: src/main/java/com/github/lunatrius/schematica/client/renderer/chunk/overlay/RenderOverlay.java (hunk 1 above)
Comment (Pokechu22) [inline src/main/java/com/github/lunatrius/schematica/client/renderer/chunk/overlay/RenderOverlay.java]: More or less (it just skips the code in some cases due to the way the `assert` keyword works).

Comment (Pokechu22) [inline src/main/java/com/github/lunatrius/schematica/client/renderer/chunk/overlay/RenderOverlay.java]: Changed to `IndexOutOfBoundsException` with a detail message.

File: src/main/java/com/github/lunatrius/schematica/client/renderer/chunk/overlay/RenderOverlay.java (hunk 2 above)
Comment (Pokechu22) [inline src/main/java/com/github/lunatrius/schematica/client/renderer/chunk/overlay/RenderOverlay.java]: Removed.

**Comment by Lunatrius** [comment]
It looks like you kept the wrong commit message when rebasing.

Other than that, did some quick tests and it seems to work just fine.

Keeping the cache inside the renderer itself is not a bad idea either, the chunk has to be refreshed anyway and keeping the old data is pointless.

**Comment by Pokechu22** [comment]
Thanks, message changed.

**Comment by Lunatrius** [comment]
Thanks!

**Pull request merged**